packages/SwingSet/src/kernel/kernel.js

```diff
@@ -289,7 +289,10 @@ export default function buildKernel(
 
       // Reject all promises decided by the vat, making sure to capture the list
       // of kpids before that data is deleted.
-      const deadPromises = [...kernelKeeper.enumeratePromisesByDecider(vatID)];
+      const deadPromises = [];
+      for (const [kpid, _p] of kernelKeeper.enumeratePromisesByDecider(vatID)) {
+        deadPromises.push(kpid);
+      }
```

Comment:
The new iterator helpers can't come soon enough. It'd be 

```js
      const deadPromises = kernelKeeper.enumeratePromisesByDecider(vatID).map(([kpid, _p]) => kpid).toArray()
```

But since you don't really need to iterate more than once, why not:

```suggestion
      const deadPromiseEntries = kernelKeeper.enumeratePromisesByDecider(vatID);
```

and later

```js
      for (const [kpid, _p] of deadPromiseEntries) {
```

Reply:
interesting.. I try to avoid keeping iterators around for very long, but that might be leftover scar tissue from when kernelKeeper's iterators mapped directly to LMDB or SQLite iterators, which didn't work well if you wanted to make other DB transactions while the iterator was held open. But `enumeratePromisesByDecider` uses `enumeratePrefixedKeys` which does a separate `kvStore.getNextKey()` for each loop, so there are no txns being held open.

The only reason I can think to not use your pattern is memory pressure: by dropping everything but the `kpid` early, we shrink the active memory footprint earlier (the peak is just as high, but it is narrower). If there is significant memory usage between the array-ification and the subsequent loop, then pruning early could help avoid extra allocation.

Reply:
actually yeah, I think I'll leave it as-is, both because of the nominal memory thing, and because in the long run I'd like to have a flavor of `enumeratePromisesByDecider` which only returns the kpids and doesn't spend any time (or DB reads) populating the rest of the data, and in that form, `deadPromises` will strictly be kpids.

Reply:
There actually should a smaller memory pressure with the iterators, since only the current entry is ever allocated, where for the array we temporarily have to allocate all entries.

Reply:
`enumeratePromisesByDecider` can be a generator function that iterates the entries and only yields the kpid. Or we can use iterator helpers' map when that is more broadly supported

Reply:
> There actually should a smaller memory pressure with the iterators, since only the current entry is ever allocated, where for the array we temporarily have to allocate all entries.

Oh that is an excellent point. Ok I think I remember the reason for the original order: the `markVatAsTerminated()` call used to be `cleanupAfterTerminatedVat()`, which deleted all the kvStore records, so we needed the list of dead promise kpids *before* everything got deleted. Now that we're merely marking it as terminated and not (yet) actually deleting everything, we can afford to fetch/iterate through the list at the end of that clause.

I'll rearrange it.